### PR TITLE
Add garbage polynomial commitment implementation for LaBRADOR protocol

### PR DESCRIPTION
**Pull request by pycckuu**
This PR implements the garbage polynomial commitment functionality for the recursive LaBRADOR protocol. The implementation handles the calculation, decomposition, and commitment of the garbage polynomials ($$g_{ij}$$ and $$h_{ij}$$) that arise from the inner product relations in the protocol.

Issue-URL:  https://github.com/example/labrador/issues/41

## How to reproduce the problem

N/A - This is a new feature implementation that adds support for the recursive structure of the LaBRADOR protocol.

## How I tested the fix

The implementation includes comprehensive unit tests that verify:
- Correct calculation of g and h polynomials
- Proper decomposition and reconstruction of polynomials
- Correct commitment generation
- Optimal parameter calculation
- Final level optimization
- Edge case handling and error conditions

## How the reviewer should test the fix

1. Review the implementation against the LaBRADOR paper, particularly Sections 5.4 and 5.6
2. Verify the correctness of the garbage polynomial calculations
3. Check the decomposition logic and parameter optimization
4. Run the test suite to ensure all tests pass

## Other useful info

The implementation includes:

1. `GarbagePolynomialCommitment` - Main struct that handles the commitment process
2. `RecursiveCommitment` - Represents the outer commitments (ν₁, ν₂)
3. `RecursiveWitness` - Contains the witness components for the next recursion level
4. `GarbageParameters` - Configures the decomposition parameters

Key features:
- Efficient calculation of g_{ij} and h_{ij} polynomials with triangular storage
- Optimal parameter calculation based on Section 5.4 of the paper
- Final level optimization from Section 5.6 for improved efficiency
- Proper error handling for invalid parameters

Note: The implementation handles the case where division by 2 is problematic in $Z_q$ by storing h'_{ij} = 2*h_{ij} directly.

### CODE REVIEW

**Comment by pycckuu** [comment]
Note to a reviewer, I added `#[allow(clippy::as_conversions)]` to some functions to allow `as` conversions. I'm wondering if there might be a better approach than using type conversions, perhaps by implementing additional traits for u32, u64, or u128.

**Review by NiDimi — CHANGES_REQUESTED** [review]
Excellent work :fire:. Great spot of the symmetric optimisation with i<=j. One thing that I think its missing, and I am open to discussing it if I am wrong, is the section 5.6, where no outer commitment is needed and there are fewer g_{ij} and h_{ij}. I suggest using a flag `final_level` to change the behavior of `commit_recursive` when its the terminal recursion step. But you might find a better approach :smile:

**Review by v0-e — APPROVED** [review]
Looks good as always :100:
We need to sync this with the recent changes by Yue already on `main`, [e.g.](https://github.com/NethermindEth/condor-rs/blob/main/labrador/src/core/aggregate.rs#L433). Feel free to explore what is best here or in another PR :)

**Comment by pycckuu** [comment]
> Excellent work 🔥. Great spot of the symmetric optimisation with i<=j. One thing that I think its missing, and I am open to discussing it if I am wrong, is the section 5.6, where no outer commitment is needed and there are fewer g_{ij} and h_{ij}. I suggest using a flag `final_level` to change the behavior of `commit_recursive` when its the terminal recursion step. But you might find a better approach 😄

Thanks @NiDimi!

This optimization is already implemented in the function `optimize_final_level`.

This function calculates the necessary components for the final level check:
1.  `g0 = Σ_i <s_i, s_i>`
2.  Generates Fiat-Shamir challenges `c_i`.
3.  Calculates the selected `g` terms (like `g_{2i-1} = <s_{2i-2}, c_i * s_{2i-1}>`) and `h` terms (like `h_{2i-1} = <φ_{2i-2}, c_i * s_{2i-1}> + <φ_{2i-1}, c_i * s_{2i-2}>`).
4.  It returns `(g0, g_terms, h_terms)`, which are the polynomials needed for the final verification equation, instead of the outer commitments `ν₁` and `ν₂`.

Instead of adding a flag to `commit_recursive`, the intended design is that the higher-level code managing the recursion calls `commit_recursive` for all intermediate levels and then explicitly calls `optimize_final_level` for the very last step. This keeps the logic for the standard recursive step and the optimized final step separate.

Does this explanation align with your understanding, or do you see a reason why integrating it into `commit_recursive` might still be preferable?

File: labrador/src/commitments/garbage_commitment.rs [hunk 1]
```diff
@@ -0,0 +1,877 @@
...
+        // Decompose using the appropriate base and number of parts
+        rq_poly.decompose(params.base(), params.num_parts())
+    }
+
+    /// Decompose all garbage polynomials into their parts
+    fn decompose_all(
+        &self,
+        g_polys: &[PolyRing],
+        h_polys: &[PolyRing],
+    ) -> (Vec<Rq<D>>, Vec<Rq<D>>) {
```
Comment (omibo) [inline labrador/src/commitments/garbage_commitment.rs:143]: g_poly and h_poly are vectors of Rq elements. The function returns a Vec<Rq>, but shouldn’t it return a Vec<Vec<Rq>> instead?

File: labrador/src/commitments/garbage_commitment.rs [hunk 2]
```diff
@@ -0,0 +1,877 @@
...
+}
+
+/// Manages garbage polynomial calculation and commitment
+pub struct GarbagePolynomialCommitment<const M: usize, const N: usize, const D: usize> {
+    // Parameters for g polynomial decomposition
+    g_decomp_params: DecompositionParameters,
+    // Parameters for h polynomial decomposition
+    h_decomp_params: DecompositionParameters,
+    // Commitment matrix for combined (t,g) polynomials as ν₁
+    nu1_matrix: RqMatrix<M, N, D>,
```
Comment (omibo) [inline labrador/src/commitments/garbage_commitment.rs:55]: In Figure 2 of the paper, it seems like B_ik, the commitment matrix for t_i^(k), isn’t just a single matrix, but rather there are r*t_1 such matrices.

**Review by omibo — CHANGES_REQUESTED** [review]

File: labrador/src/commitments/garbage_commitment.rs [hunk 3]
```diff
@@ -0,0 +1,877 @@
...
+        // 4. Create witnesses for outer commitments
+        let nu1_witness = self.create_witness_from_parts(&combined_parts);
+        let nu2_witness = self.create_witness_from_parts(&h_parts);
+
+        // 5. Create outer commitments (ν₁, ν₂)
+        let nu1_commitment = AjtaiCommitment::new(self.params.clone(), self.nu1_matrix.clone())?;
+        let (nu1, _) = nu1_commitment.commit(nu1_witness)?;
+
+        let nu2_commitment = AjtaiCommitment::new(self.params.clone(), self.nu2_matrix.clone())?;
+        let (nu2, _) = nu2_commitment.commit(nu2_witness)?;
```
[the remark this answers is not in the document]
Comment (v0-e) [reply labrador/src/commitments/garbage_commitment.rs:225]: To be handled in #60

**Review by v0-e — APPROVED** [review]

**Review by NiDimi — APPROVED** [review]
LGTM!

**Review by omibo — APPROVED** [review]
